### PR TITLE
Dynamic mock client (2)

### PR DESCRIPTION
**Pull request by johnstonematt**

### CODE REVIEW

File: cmd/solana_exporter/dynamic_test.go
Comment (SEJeff) [inline cmd/solana_exporter/dynamic_test.go:1]: There is no strict rule necessarily to keep the test files the same names as the filenames, but I have literally never seen a project in go that did this and I've shipped more go to production than python. It is very strange to see.

File: cmd/solana_exporter/exporter_test.go
Comment (SEJeff) [inline cmd/solana_exporter/exporter_test.go:1]: boo. General idiom is that the test files match the source files.

File: cmd/solana_exporter/slots.go [hunk 1]
```diff
@@ -65,10 +65,10 @@ func init() {
 	prometheus.MustRegister(leaderSlotsByEpoch)
 }
 
-func (c *solanaCollector) WatchSlots() {
+func (c *solanaCollector) WatchSlots(ctx context.Context) {
 	// Get current slot height and epoch info
-	ctx, cancel := context.WithTimeout(context.Background(), httpTimeout)
-	info, err := c.rpcClient.GetEpochInfo(ctx, rpc.CommitmentMax)
+	ctx_, cancel := context.WithTimeout(context.Background(), httpTimeout)
+	info, err := c.rpcClient.GetEpochInfo(ctx_, rpc.CommitmentMax)
```
Comment (SEJeff) [inline cmd/solana_exporter/slots.go:71]: Why are you creating two contexts here vs just using the one you pass in? I don't quite follow.

File: cmd/solana_exporter/slots.go [hunk 2]
```diff
@@ -89,44 +89,78 @@ func (c *solanaCollector) WatchSlots() {
...
-		if err != nil {
-			klog.Infof("failed to fetch epoch info, retrying: %v", err)
+		default:
+			<-ticker.C
+
+			// Get current slot height and epoch info
+			ctx_, cancel := context.WithTimeout(context.Background(), httpTimeout)
+			info, err := c.rpcClient.GetEpochInfo(ctx_, rpc.CommitmentMax)
+			if err != nil {
+				klog.Infof("failed to fetch epoch info, retrying: %v", err)
```
Comment (SEJeff) [inline cmd/solana_exporter/slots.go:107]: I realize you're just moving the code around, but consider making this a warning as an error isn't really an info.

File: cmd/solana_exporter/slots.go [hunk 3]
```diff
@@ -89,44 +89,78 @@ func (c *solanaCollector) WatchSlots() {
...
-			continue
-		}
-		cancel()
 
-		if watermark == info.AbsoluteSlot {
-			klog.Infof("slot has not advanced at %d, skipping", info.AbsoluteSlot)
-			continue
-		}
+			if watermark == info.AbsoluteSlot {
+				klog.Infof("slot has not advanced at %d, skipping", info.AbsoluteSlot)
```
Comment (SEJeff) [inline cmd/solana_exporter/slots.go:114]: debug

File: cmd/solana_exporter/static_test.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	// reset metrics before running tests:
+	leaderSlotsTotal.Reset()
+	leaderSlotsByEpoch.Reset()
+
+	collector := createSolanaCollector(
+		&staticRPCClient{},
+		100*time.Millisecond,
+	)
+	prometheus.NewPedanticRegistry().MustRegister(collector)
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (SEJeff) [inline cmd/solana_exporter/static_test.go:93]:
```suggestion
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
```

A bit easier to just defer the cancel in places like this vs some of the places in the code where you have it sprinkled all over some gnarly logic.

File: cmd/solana_exporter/static_test.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+	go collector.WatchSlots(ctx)
+	time.Sleep(1 * time.Second)
```
Comment (SEJeff) [inline cmd/solana_exporter/static_test.go:95]: A waitgroup (from the sync package) is probably too much for tests, but is this necessary? Generally speaking, I see most code like this as code smell.

File: cmd/solana_exporter/static_test.go [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+					for _, identity := range identities {
+						testBlockProductionMetric(t, metric, identity, status)
+					}
+				})
+			}
+		})
+	}
+	// cancel and wait for cancellation:
+	cancel()
+	time.Sleep(time.Second)
```
Comment (SEJeff) [inline cmd/solana_exporter/static_test.go:137]:
```suggestion
```

File: go.mod [hunk 7]
```diff
@@ -1,10 +1,21 @@
 module github.com/certusone/solana_exporter
 
-go 1.13
+go 1.22
```
Comment (SEJeff) [inline go.mod:3]: Just pull this in from your rebase.

**Review by SEJeff — APPROVED** [review]

File: cmd/solana_exporter/dynamic_test.go [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+		// make current final the new initial (for next iteration)
+		initial = final
+	}
+
+	// epoch should have changed somewhere
+	assert.Truef(t, epochChanged, "Epoch has not changed!")
+
+	// cancel and wait for cancellation:
+	slotsCancel()
+	runCancel()
```
[the remark this answers is not in the document]
Comment (johnstonematt) [reply cmd/solana_exporter/dynamic_test.go:218]: will change

File: cmd/solana_exporter/slots.go (hunk 2 above)
Comment (johnstonematt) [reply cmd/solana_exporter/slots.go:107]: will do

File: cmd/solana_exporter/static_test.go (hunk 5 above)
Comment (johnstonematt) [reply cmd/solana_exporter/static_test.go:95]: It is to let the exporter run for a bit before we test it, to make sure it has gathered metrics before we go look what those metrics are and test them.

File: go.mod (hunk 7 above)
Comment (johnstonematt) [reply go.mod:3]: The Go Update PR was to merge that branch into this branch

**Pull request merged**